src/sql/src/vars.rs

```diff
@@ -17,16 +17,49 @@ use once_cell::sync::Lazy;
 use serde::Serialize;
 use uncased::UncasedStr;
 
+use crate::ast::Ident;
+use crate::DEFAULT_SCHEMA;
 use mz_build_info::BuildInfo;
 use mz_ore::cast;
 use mz_persist_client::cfg::PersistConfig;
-use mz_sql::ast::Ident;
-use mz_sql::DEFAULT_SCHEMA;
 use mz_sql_parser::ast::TransactionIsolationLevel;
 
-use crate::catalog::SYSTEM_USER;
-use crate::error::AdapterError;
-use crate::session::{EndTransactionAction, User};
+/// The action to take during end_transaction.
+///
+/// This enum lives here because of convenience: it's more of an adapter
+/// concept but [`SessionVars::end_transaction`] takes it.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum EndTransactionAction {
+    /// Commit the transaction.
+    Commit,
+    /// Rollback the transaction.
+    Rollback,
+}
+
+/// Errors that can occur when working with [`Var`]s
+#[derive(Debug, thiserror::Error)]
+pub enum VarError {
+    #[error("unknown parameter: {0}")]
+    UnknownParameter(String),
+    #[error("value for parameter {0:?} has wrong type")]
+    InvalidParameterType(&'static (dyn Var + Send + Sync)),
+    #[error("value for parameter {parameter:?}: {values:?} is wrong because of {reason}")]
+    InvalidParameterValue {
+        parameter: &'static (dyn Var + Send + Sync),
+        values: Vec<String>,
+        reason: String,
+    },
+    #[error("specified parameter {0:?} is fixed to a specific value")]
+    FixedValueParameter(&'static (dyn Var + Send + Sync)),
+    #[error("specified parameter {0:?} is read only")]
+    ReadOnlyParameter(&'static (dyn Var + Send + Sync)),
+    #[error("value for parameter {parameter:?}: {values:?} is constrainted to {valid_values:?}")]
+    ConstrainedParameter {
+        parameter: &'static (dyn Var + Send + Sync),
+        values: Vec<String>,
+        valid_values: Option<Vec<&'static str>>,
+    },
+}
```

Comment:
When splitting out an error enum like this, it's best to make the new enum _the_ definition of the variants, rather than leaving the variants in the original enum. Avoids having two copies of the variants, and all the boilerplate in the `From` impl. Here's what I mean: https://github.com/MaterializeInc/materialize/pull/17961/commits/708f4a2812757dec48b8bee9c7a475d0509b9bf5(Noticed this while rebasing that PR.)

Looks good otherwise, though—thanks for slogging! These kinds of refactors are never fun.

Reply:
Thanks. Yeah that diff looks better